### ISSUE TITLE
Support counts on `m`/`M`

### ISSUE THREAD

**Issue by lenormf** [mawww/kakoune#2097]
Hi,

Small discussion on IRC:

>ironzorg | consider this: `%(i)nt fn() {` to select the entire function, I could do `2M`
>mawww_imc | how about `%(i)nt fn(int a = fn2()) {` ? would 2M select till `fn2()` or till closing `}` ?

**Comment by Delapouite** [comment]
Somehow related: [#795 Allow counts for text object start/end selections](https://github.com/mawww/kakoune/issues/795)

**Comment by lenormf** [comment]
I'm thinking that [@mawww's suggestion](https://github.com/mawww/kakoune/issues/795#issuecomment-249048818) in #795 would be best.

```
%(d)ef foo(bar=[1, 2, 3]):
```

<kbd>2</kbd><kbd>m</kbd>:

```
[1, 2, 3]
```

<kbd>2</kbd><kbd>M</kbd>:

```
def foo(bar=[1, 2, 3]
```

**Comment by Delapouite** [comment]
Yes this behavior seems useful.

How would we formalize this rule in plain English?
Should we talk about nesting, aka parent/child relationship. Or in terms of "skipping `count` matching chars".

Also what would `2m` do if the cursor is inside the square brackets?
`def foo(bar=[%(1), 2, 3]):`

I guess it would select the surrounding parens:

`(bar=[1, 2, 3])`

In the end having count doing "something" with `m` is always better than doing nothing. So we can go this way as an experiment and adjust the behavior after feedback from several users if needed.

I think some insights from developers used to working with LISP-like languages and emacs commands like @eraserhd would be valuable.

**Comment by lenormf** [comment]
>Or in terms of "skipping count matching chars".

I think so, which means your example is correct, the parenthesis would be selected with this rule.

**Comment by eraserhd** [comment]
I think the way to describe this in English is that `m` scans forward for the `count`'th bracket, then highlights its match.

I don't think it really fits into Lisp navigation, but it seems general enough to be useful.  In the case where there's a line like:
```
%(x) (one) (two) (three)
```

So 1m highlights `(one)`, 2m also highlights `(one)`, 3m highlights `(two)`... and so forth.

Making it work in terms of parenthesized regions instead of bracket count is harder and harder to explain, because now we have to deal with the fact that parentheses form a tree, and have to move among siblings, children, and ancestors of the objects.  We'd need more commands than just one (next/previous sibling, next bigger brackets, and, say, first sibling of selection).

**Comment by lenormf** [comment]
>I don't think it really fits into Lisp navigation, but it seems general enough to be useful. In the case where there's a line like:
>%(x) (one) (two) (three)
>So 1m highlights (one), 2m also highlights (one), 3m highlights (two)... and so forth.

I realised that this use-case would be easy to implement with a mapping: just execute <kbd>l</kbd> or <kbd>L</kbd> after each <kbd>m</kbd> or <kbd>M</kbd>.

Making the counts refer to the level of nesting seems harder, and I've met opportunities of use in my workflow, so it's probably better to have that one built-in.
